golem/resources/debit_note/debit_note.py

```diff
@@ -45,58 +45,92 @@ def activity(self) -> "Activity":
     async def get_status(self) -> PayDocumentStatus:
         return PayDocumentStatus(str((await self.get_data()).status).upper())
 
-    async def get_previous_debit_note_data(self, timestamp: datetime) -> Optional[models.DebitNote]:
-        """Get latest debit note data from before given `timestamp`."""
-        debit_notes_data = [(await dn.get_data()) for dn in self.parent.debit_notes]
-        debit_notes_data.sort(key=lambda dn: dn.timestamp, reverse=True)
-        for dn in debit_notes_data:
-            if dn.timestamp < timestamp:
-                return dn
+    async def get_previous_debit_note(self) -> Optional["DebitNote"]:
+        """Get previous debit note."""
+        for debit_note in sorted(
+            self.activity.debit_notes, key=lambda dn: dn.created_at, reverse=True
+        ):
+            if debit_note.created_at < self.created_at:
+                return debit_note
         return None
 
-    async def get_previous_debit_notes_count(self, timestamp: datetime) -> int:
-        """Get debit note count from before given `timestamp`."""
-        debit_notes_data = [(await dn.get_data()) for dn in self.parent.debit_notes]
-        return len([dn for dn in debit_notes_data if dn.timestamp < timestamp])
+    async def get_previous_payable_debit_note(self) -> Optional["DebitNote"]:
+        """Get previous payable debit note."""
+        for debit_note in sorted(
+            self.activity.debit_notes, key=lambda dn: dn.created_at, reverse=True
+        ):
+            await debit_note.get_data()  # ensure debit_note.data is accessible
+            if (
+                debit_note.created_at < self.created_at
+                and debit_note.data.payment_due_date is not None
+            ):
+                return debit_note
+        return None
```

Comment:
again, we can do away with the sorting if it's just one value that we're after:


```

        try:
            return max(
                [
                    dn for dn in self.activity.debit_notes
                    if dn.created_at > self.created_at and (await dn.get_data()).payment_due_date is not None
                ],
                key=lambda dn: dn.created_at,
            )
        except ValueError:
            return None
```

plus, it's more natural to just _use_ the value returned by `get_data()` instead of adding the "ensure..." comment...

Reply:
Turns out it throws `async_generator` object is not iterable